index.html

```diff
@@ -140,76 +190,254 @@ <h2>
     </section>
-    <section data-dfn-for="Badge">
+    <section>
       <h2>
-        <code>Badge</code> namespace
+        Extensions to the `Navigator` and `WorkerNavigator` interfaces
       </h2>
+      <p>
+        The {{Navigator}} and {{WorkerNavigator}} interfaces are extended with
+        methods for setting and clearing both the document and application
+        badge indicators from documents and service workers, respectively.
+      </p>
       <pre class="idl">
-        [Exposed=(Window)]
-        namespace Badge {
-          void set(optional [EnforceRange] unsigned long long contents);
-          void clear();
+        // Methods only exposed on documents.
+        [SecureContext]
+        partial interface Navigator {
+          Promise&lt;void&gt; setClientBadge(optional [EnforceRange] unsigned long long contents);
+          Promise&lt;void&gt; clearClientBadge();
+        };
+
+        // Methods exposed on both documents and service workers.
+        [SecureContext]
+        interface mixin NavigatorBadge {
+          Promise&lt;void&gt; setAppBadge(optional [EnforceRange] unsigned long long contents);
+          Promise&lt;void&gt; clearAppBadge();
         };
+
+        Navigator includes NavigatorBadge;
+        WorkerNavigator includes NavigatorBadge;
       </pre>
       <p>
-        The <dfn>Badge</dfn> namespace provides functions for setting and
-        clearing the application badge indicator.
+        User agents that never display document badges SHOULD NOT expose the
+        {{Navigator/setClientBadge()}} and {{Navigator/clearClientBadge()}}
+        methods. Similarly, user agents that never display application badges
+        SHOULD NOT expose the {{NavigatorBadge/setAppBadge()}} and
+        {{NavigatorBadge/clearAppBadge()}} methods.
+      </p>
+      <p class="note">
+        This is important as a feature-detection mechanism to allow sites to
+        determine whether setting a badge will have any effect, based on the
+        presence of the methods. In particular, sites can use the absence of
+        {{Navigator/setClientBadge()}} as a condition for displaying a fallback
+        badge in the page's [^title^] or <a data-cite=
+        "html#rel-icon"><code>icon</code></a>.
       </p>
+      <div class="issue" data-number="60"></div>
+      <div class="issue" data-number="61"></div>
       <section>
         <h3>
-          <dfn>set()</dfn> function
+          <dfn data-dfn-for="Navigator">setClientBadge()</dfn> method
         </h3>
         <p>
-          When the <code>set()</code> function is called with argument
-          <var>contents</var>, run the following steps:
+          When the {{Navigator/setClientBadge()}} method is called with
+          argument |contents:unsigned long long|:
         </p>
         <ol>
-          <li>If <var>contents</var> is omitted, set the application's badge to
-          <a>flag</a>.
+          <li>Return [=a promise resolved with=] undefined, and [=in
```

Comment:
This is still not correct as it could come back to bite us. This should run in parallel:

1. request set the badge, 
1. possibly reject with an error (let's just add a note saying that we are still gathering implementation experience on what errors could results on different platform/implementations... it's an incubation after all)
1. If no error, resolve the promise with undefined.